sound/soc/sof/topology.c

```diff
@@ -1996,12 +2000,28 @@ static int sof_manifest(struct snd_soc_component *scomp, int index,
 		}
 	}
 
-	/* copy nhlt blob after that should be after ABI data */
+	/* we have more data in manifest than just abi number, so parse it */
 	if (size > SOF_TPLG_ABI_SIZE) {
-		sdev->nhlt_blob = devm_kmemdup(sdev->dev, &man->priv.data[SOF_TPLG_ABI_SIZE],
-					       size - SOF_TPLG_ABI_SIZE, GFP_KERNEL);
-		if (!sdev->nhlt_blob)
-			return -ENOMEM;
+		manifest = (struct sof_manifest *)man->priv.data;
+		manifest_data = manifest->items;
+		len_check = SOF_TPLG_ABI_SIZE + sizeof(manifest->count);
+		for (i = 0; i < manifest->count; i++) {
+			len_check += sizeof(struct sof_manifest_data) + manifest_data->size;
+			if (len_check > size)
+				return -EINVAL;
+			switch (manifest_data->type) {
+			case SOF_MANIFEST_DATA_TYPE_NHLT:
+				sdev->nhlt_blob = devm_kmemdup(sdev->dev, manifest_data->data,
+							       manifest_data->size, GFP_KERNEL);
+				if (!sdev->nhlt_blob)
+					return -ENOMEM;
+				break;
+			default:
+				return -EINVAL;
+			}
+			manifest_data = (struct sof_manifest_data *)
+			  (sizeof(struct sof_manifest_data) + manifest_data->size);
```

Comment:
you sure about this one? you convert a sizeof + size in a pointer?

Reply:
so jumping in a memory area of variable size chunks... so go forward size of the "header" + data following it, and cast that address into another header (in a loop). The size doesn't contain the size of the struct itself. But yeah, any suggestions then?

Reply:
tried to make this better now...